#### Detecting fake news using neive bayse classifier

In [32]:
import os
os.chdir('C://Users/Ajinkya/Dropbox/Spring 2018/BUAN 6340/Project/NLP_project')

In [33]:
# importing required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [34]:
# loading dataset
news = pd.read_csv('fake_or_real_news.csv', index_col=0)

In [35]:
news.head()

title  \
8476                        You Can Smell Hillary’s Fear   
10294  Watch The Exact Moment Paul Ryan Committed Pol...   
3608         Kerry to go to Paris in gesture of sympathy   
10142  Bernie supporters on Twitter erupt in anger ag...   
875     The Battle of New York: Why This Primary Matters   

                                                    text label  
8476   Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
10294  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
3608   U.S. Secretary of State John F. Kerry said Mon...  REAL  
10142  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
875    It's primary day in New York and front-runners...  REAL

In [36]:
news.shape

(6335, 3)

In [37]:
# counting number of fake and real news in dataset
news.label.value_counts()

REAL    3171
FAKE    3164
Name: label, dtype: int64

##### Dataset looks balanced

In [38]:
# importing libraries
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

In [39]:
X = news['text']
y = news['label']

In [40]:
# splitting datset into train and test dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 100)

In [41]:
y_train.value_counts()

REAL    2220
FAKE    2214
Name: label, dtype: int64

In [42]:
y_test.value_counts()

REAL    951
FAKE    950
Name: label, dtype: int64

##### test datset is well balanced

In [43]:
# creating countVectorizer object and fitting training dataset on it.
# removed stop-words in dataset
count = CountVectorizer(stop_words='english')
news_train_count = count.fit_transform(X_train)
news_test_count = count.transform(X_test)

In [44]:
news_train_count.shape

(4434, 58197)

In [45]:
# neive bayes classifier
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(news_train_count, y_train)

In [46]:
predicted = clf.predict(news_train_count)
np.mean(predicted == y_train)

0.94091114118177721

In [47]:
# calculating test accuracy
predicted = clf.predict(news_test_count)
np.mean(predicted == y_test)

0.88374539715938982

In [48]:
from sklearn.metrics import classification_report
confusion_matrix = classification_report(y_test,predicted)
print(confusion_matrix)

             precision    recall  f1-score   support

       FAKE       0.90      0.86      0.88       950
       REAL       0.87      0.91      0.89       951

avg / total       0.88      0.88      0.88      1901



##### able to achieve overall 88% accuracy on test datset

##### Using TfIdf

In [49]:
# creating TfIdfVectorizer object and fitting data
count_tfidf = TfidfVectorizer(stop_words='english')
news_train_tfidf = count_tfidf.fit_transform(X_train)
news_test_tfidf = count_tfidf.transform(X_test)

In [50]:
# neive bayes
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(news_train_tfidf, y_train)

In [51]:
predicted = clf.predict(news_train_tfidf)
np.mean(predicted == y_train)

0.90099233198015338

In [52]:
# predicting accuracy
predicted = clf.predict(news_test_tfidf)
np.mean(predicted == y_test)

0.84166228300894264

In [53]:
confusion_matrix_tfidf = classification_report(y_test,predicted)
print(confusion_matrix_tfidf)

             precision    recall  f1-score   support

       FAKE       0.97      0.71      0.82       950
       REAL       0.77      0.97      0.86       951

avg / total       0.87      0.84      0.84      1901



##### achieved accuracy of 87%

#### using CountVectorizer we able to detect almost same percent of fake and real news. But using tfidfVectoriser detection accuracy of real news decreases to 77%